nav2_costmap_2d/src/costmap_layer.cpp

```diff
@@ -86,6 +86,11 @@ void CostmapLayer::updateWithTrueOverwrite(
   if (!enabled_) {
     return;
   }
+
+  if (costmap_ == nullptr) {
+    throw "Can't update costmap layer. It has't been initilized yet. Check if /map topic is published.";
```

Comment:
We should throw an actual exception class. Not just a string. eg. `throw std::runtime_error("...");`

Bonus points if you define a custom exception for this specific error.

Reply:
I don't think this should be an exception at all, look above it when we just check if its a valid state to do the thing, we just check if `enabled_` is valid and return if not

`if (!enabled_ || !costmap_) {return;}` type logic. There's not exceptions anywhere in costmap_2d, that I can recall on hand, I'd like to not add them.

Reply:
costmap_ == nullptr is a fatal programming error IMHO. We should terminate execution. Throwing an exception is the best way to do that

Reply:
To expand on that, if you don't want to do an update, you should have enabled==false. You should never be in a situation where enabled==true and costamp_==false. That's an undefined system state

Reply:
I agree that it shouldn't happen, but when you introduce the exception if not caught it will bring down the system. I'd rather it pass an update in a bad state than go down

Reply:
Why would you want to operate in an invalid state? Who knows what bad things could happen? It'd be much better to fail immediately so, whoever is writing this layer plugin can fix the problem early on.

Reply:
Except when that weird issue happens in a different setting where bringing down costs money, I'd rather spam with errors and continue to attempt to work than to just stop in the middle of a warehouse indefinitely. 

Also I'm remembering that this check wasn't there prior, if this is an invalid state and it hasn't been an issue for the last 8 years or so, why not just get rid of it?

Reply:
In the past, this used to seg fault. It didn't continue on. That's what alerted us that there was a race condition in the static layer.

Masking this problem means you are not getting costmap updates when you think you should be. That could potentially lead to not updating the costmap with obstacle data and running into things. I feel that is not a situation we'd want to continue operation. 

However, by throwing the exception, higher levels of the stack could decide to soldier on or to try some sort of recovery, such as reinitializing the costmap. I'd be OK with seg faulting, but the exception gives us the potential for more intelligent error handling

Reply:
Sure, but then we need to add exception handling anywhere touching a layer update. Which I don't think is bad just more to scope into the PR before merging

Reply:
>  I'd rather spam with errors and continue to attempt to work than to just stop in the middle of a warehouse indefinitely

We should probably have either a compile option or a parameter that lets you choose intolerant error handling for development and tolerant error handling for production.

Reply:
> Ok I can get behind that, but then we need to add exception handling anywhere touching a layer update.

@mlherd So, in the costmap_layer.cpp, there are a number of "update*" functions that all dereference costmap. We should add the nullptr check in each.

Reply:
Not adding those checks here, out of scope for this PR. Please file an issue if that should be done in the future.